File: frontend/app/(cover-letter-app)/api/cover-letter.ts
import { getAuthFromStorage } from '@/lib/sessionStorage';
import { authenticatedRequest } from '@/lib/apiClient';

const API_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000/api';

// Define types based on our backend schema
export interface CoverLetterBase {
  company_name: string;
  hiring_manager?: string | null;
  job_description: string;
  generated_content?: string | null;
  status?: string;
}

export interface CoverLetterCreate extends CoverLetterBase {
  // No additional fields needed
}

export interface CoverLetterUpdate {
  company_name?: string;
  hiring_manager?: string | null;
  job_description?: string;
  generated_content?: string | null;
  status?: string;
}

export interface CoverLetterExperienceLink {
  experience_id: string;
  relevance_order: number;
}

export interface CoverLetter extends CoverLetterBase {
  id: string; // UUID as string
  user_id: string; // UUID as string
  created_at?: string; // ISO format datetime string
  updated_at?: string; // ISO format datetime string
  selected_experiences?: CoverLetterExperienceLink[];
}

// Create a new cover letter
export const createCoverLetter = async (coverLetter: CoverLetterCreate): Promise<CoverLetter> => {
  try {
    return await authenticatedRequest<CoverLetter>('cover-letters', {
      method: 'POST',
      body: JSON.stringify(coverLetter)
    });
  } catch (error) {
    console.error("Error creating cover letter:", error);
    throw error;
  }
};

// Get all cover letters for the current user
export const getUserCoverLetters = async (): Promise<CoverLetter[]> => {
  try {
    return await authenticatedRequest<CoverLetter[]>('cover-letters', {
      method: 'GET'
    });
  } catch (error) {
    console.error("Error fetching cover letters:", error);
    throw error;
  }
};

// Get a specific cover letter by ID
export const getCoverLetter = async (coverLetterId: string): Promise<CoverLetter> => {
  try {
    return await authenticatedRequest<CoverLetter>(`cover-letters/${coverLetterId}`, {
      method: 'GET'
    });
  } catch (error) {
    console.error(`Error fetching cover letter ${coverLetterId}:`, error);
    throw error;
  }
};

// Update an existing cover letter
export const updateCoverLetter = async (coverLetterId: string, update: CoverLetterUpdate): Promise<CoverLetter> => {
  try {
    return await authenticatedRequest<CoverLetter>(`cover-letters/${coverLetterId}`, {
      method: 'PUT',
      body: JSON.stringify(update)
    });
  } catch (error) {
    console.error(`Error updating cover letter ${coverLetterId}:`, error);
    throw error;
  }
};

// Delete a cover letter
export const deleteCoverLetter = async (coverLetterId: string): Promise<void> => {
  try {
    await authenticatedRequest(`cover-letters/${coverLetterId}`, {
      method: 'DELETE'
    });
  } catch (error) {
    console.error(`Error deleting cover letter ${coverLetterId}:`, error);
    throw error;
  }
};

// Generate cover letter content using AI
export interface GenerateCoverLetterRequest {
  company_name: string;
  hiring_manager?: string;
  job_description: string;
  experiences: Array<{
    title: string;
    description: string;
    skills: string[];
    duration: string;
  }>;
}

export interface GenerateCoverLetterResponse {
  cover_letter: string;
  chances: string;
  chances_explanation: string;
}

export const generateCoverLetterContent = async (request: GenerateCoverLetterRequest): Promise<GenerateCoverLetterResponse> => {
  try {
    return await authenticatedRequest<GenerateCoverLetterResponse>('cover-letters/generate', {
      method: 'POST',
      body: JSON.stringify(request)
    });
  } catch (error) {
    console.error("Error generating cover letter content:", error);
    throw error;
  }
};

// Add an experience to a cover letter
export const addExperienceToCoverLetter = async (
  coverLetterId: string, 
  experienceId: string, 
  relevanceOrder: number
): Promise<any> => {
  try {
    return await authenticatedRequest(`cover-letters/${coverLetterId}/experiences/${experienceId}?relevance_order=${relevanceOrder}`, {
      method: 'POST'
    });
  } catch (error) {
    console.error(`Error adding experience to cover letter:`, error);
    throw error;
  }
};

// Remove an experience from a cover letter
export const removeExperienceFromCoverLetter = async (
  coverLetterId: string, 
  experienceId: string
): Promise<void> => {
  try {
    await authenticatedRequest(`cover-letters/${coverLetterId}/experiences/${experienceId}`, {
      method: 'DELETE'
    });
  } catch (error) {
    console.error(`Error removing experience from cover letter:`, error);
    throw error;
  }
};

File: backend/schemas/cover_letter.py
from pydantic import BaseModel, UUID4, Field
from typing import List, Optional
from datetime import datetime
from uuid import UUID


class CoverLetterBase(BaseModel):
    """Base schema for cover letter data."""
    job_title: str = Field(..., description="The title of the job being applied for")
    company_name: str = Field(..., description="The name of the company")
    job_description: str = Field(..., description="The full job description")
    tone: str = Field(..., description="The desired tone of the cover letter (e.g., professional, friendly)")
    max_length: int = Field(..., description="Maximum length of the cover letter in words", ge=100, le=1000)


class CoverLetterCreate(CoverLetterBase):
    """Schema for creating a new cover letter."""
    pass


class CoverLetterUpdate(BaseModel):
    """Schema for updating an existing cover letter."""
    job_title: Optional[str] = Field(None, description="The title of the job being applied for")
    company_name: Optional[str] = Field(None, description="The name of the company")
    job_description: Optional[str] = Field(None, description="The full job description")
    tone: Optional[str] = Field(None, description="The desired tone of the cover letter")
    max_length: Optional[int] = Field(None, description="Maximum length of the cover letter in words", ge=100, le=1000)


class CoverLetterExperienceLink(BaseModel):
    experience_id: UUID4
    relevance_order: int


class CoverLetter(CoverLetterBase):
    id: UUID4
    user_id: UUID4
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    selected_experiences: Optional[List[CoverLetterExperienceLink]] = None
    
    class Config:
        from_attributes = True


class ExperienceInCoverLetter(BaseModel):
    """Schema for an experience included in a cover letter."""
    id: UUID
    title: str
    company: str
    start_date: datetime
    end_date: Optional[datetime]
    description: str
    key_achievements: List[str]


class CoverLetterOutput(CoverLetterBase):
    """Schema for cover letter response data."""
    id: UUID
    user_id: UUID
    content: str
    created_at: datetime
    updated_at: datetime
    experiences: List[ExperienceInCoverLetter] = Field(default_factory=list)

    class Config:
        from_attributes = True


class CoverLetterInDB(CoverLetterOutput):
    pass 